src/newscale/stages.py

```diff
@@ -1,365 +1,497 @@
+                    and stats[x][StateBit.ON_TARGET]
+                    for x in axes.keys()]):
+                return
+            sleep(0.01)
+        raise RuntimeError("Axes timed out trying to reach target position.")
+
+    @axis_check('wait')
+    def move_for_time(self, wait: bool = True,
+                      **axes: Tuple[Direction, Tuple[float, None]]):
+        """Move axes specified for the corresponding amount of time in seconds.
+
+        :param wait: bool indicating if this function should block until the
+            stage has reached its destination.
+        :param axes: a per-axis tuple of (Direction, <move_time_in_seconds>).
+            the move time can be set to None, and the axis will run until
+            reaching a limit or being issued a halt.
+        """
+        for axis_name, (direction, seconds) in axes.items():
+            self.stages[axis_name].run(direction, seconds)
+        if not wait:
+            return
+        # Poll position vector until we have reached the target or timeout.
+        start_time = perf_counter()
+        while perf_counter() - start_time < self.__class__.MOVE_TIMEOUT_S:
+            stats = {x: self.stages[x].get_closed_loop_state_and_position()[0]
+                     for x in axes.keys()}
+            if any([stats[x][StateBit.STALLED] for x in axes.keys()]):
+                raise RuntimeError("One or more axes is stalled.")
+            if all([(not stats[x][StateBit.RUNNING])
+                    and (not stats[x][StateBit.TIMED_RUN])
+                    for x in axes.keys()]):
+                return
+            sleep(0.01)
+        raise RuntimeError("One or more axes timed out trying to move for the"
+                           "specified time.")
+
+    @axis_check()
+    def get_position(self, *axes: str):
+        """Retrieve the specified axes positions (in mm) as a dict, or get all
+        positions if none are specified.
+
+        :param axes: an unlimited number of axes specified by name (string).
+
+        .. code-block:: python
+
+            stages.get_position('x', 'y', 'z')  # Get specified positions OR
+            stages.get_position()  # Get all positions.
+
+        """
+        if len(axes) == 0:  # return all axes if None are specified.
+            axes = self.stages.keys()
+        return {x: self.stages[x].get_position() for x in axes}
+
+    @axis_check()
+    def set_speed(self, **axes):
+        """Set the speeds of the specified axes in mm/sec."""
+        # TODO: Units
+        pass
+
+    def get_speed(self, **axes):
+        """Get the speeds of the specified axes"""
+        # TODO: Units
+        pass
+
+    @axis_check()
+    def halt(self):
+        """Halt all axes"""
+        return {x: self.stages[x].halt() for x in self.stages.keys()}
 
+    def close(self):
+        """Release computer control of all axes."""
+        return {x: self.stages[x].close() for x in self.stages.keys()}
+
+
+class USBXYZStage(MultiStage):
```

Comment:
this is really nice, and I expect 99% of the time we'll just be instantiating these convenience classes. as such, i'd like to make them more obvious.

first, can they be in their own submodule, or perhaps the top-level module? i.e. "from newscale import USBXYZStage"

second, i'm not clear about the arguments? i'm guessing port would be "/dev/ttyUSB0" or similar, but how (or why) would you pass serial_itnerface? does it make sense to get rid of it?

third, related to my comment above, i think this should specify the type of interface more specifically, something like USB_XYZ_M3_LS_34_15, or whatever we end up calling the 1-axis subclass.

Reply:
Hmm, I could make it appear like:
````python
from newscale import USBXYZStage
````
But I really do think keeping it in stages makes sense. That's because I see this package not just as a way to drive these XYZ stages, but as a way to talk to all sorts of stages in general. Putting things at the stage module level
````python
from newscale.stages import M3LinearSmartStage
from newscale.stages import USBXYZStage
````
makes it clear that there are many options here. To your other comments about more detailed stage classes, that kinda further drives my point home about wanting to distinguish that there are many stages.
````python
from newscale.stages import M3LS34, M3LSU2, M3LS18
````

It's also likely that a general purpose user will also need to import
````python
from newscale.device_codes import Direction, DriveMode
````
Having that stages separation is nice and explicit.

### Args
Re arguments, I want to support the creation of a generic USB interface that may talk to more-than-3 stages. In that case, you might already have the interface object already created. In that case you can pass it in as a parameter. Example:
````python
from newscale.stages import USBXYZStage, M3LinearSmartStage
from newscale.interface import USBInterface

teensy = USBInterface("/dev/ttyUSB0") # this could be our own widget, with N spi ports supporting N stages.
extra_stage = M3LinearSmartStage(interface=teensy, address="04")
xyz_stage = USBXYZStage(interface=teensy)  # same teensy using stages on "01", "02", "03" addresses
````
The nice thing about the above is that there are no code changes at this level to implement the above sort of device.

Finally, since the above use-case is kinda niche, the most-likely situation has a simple function signature.
````python
stage = M3LinearSmartStage("/dev/ttyUSB0", address="01")
xyz_stage = USBXYZStage("/dev/ttyUSB1")
````

> third, related to my comment above, i think this should specify the type of interface more specifically, something like USB_XYZ_M3_LS_34_15, or whatever we end up calling the 1-axis subclass.
This I'm willing to explore if I can find commands that differ across stages.

Reply:
"Re arguments, I want to support the creation of a generic USB interface that may talk to more-than-3 stages."

this suggests to me that we are not on the same page about what this USBXYZStage should be. but actually i think that's a good thing - if we can get there then i think we will agree on the implementation.

from my perspective, you have chosen an abstraction architecture (stage, interface, multistage) that is not totally obvious to me, in order to solve a problem we don't currently have (mix-and-match "xyzu" mutlistages). maybe you are right and this is a good thing, so i don't want to nix it right now; i do think we should talk about it a bit more.

my suggestion is, let's keep that stuff, but then introduce a set of convenience classes which represent the most common way we (and i would think, 99% of users) will use this library. when you purchase an XYZ stage from newscale, it comes with an interface, either USB or PoE, which connects in a canonical way: x to x, y to y, z to z. let's just make a object class which represents that product.

that's what i thought USBXYZStage was. therefore it should be instantiated with a single comport/deviceFile. and i think it should live in a convenience module, either newscale.convenience or newscale.products or just newscale (top level). i don't see any reason why the 99% person should have to learn the stage/interface/multistage abstraction if they don't need to.

this approach seems good from a development, separation-of-concerns standpoint too - we can continue to explore the general stage/interface mappings at the low level, while primarily using the convenience classes in our rigs which transparently depend on the low-level.

Reply:
> that's what i thought USBXYZStage was. therefore it should be instantiated with a single comport/deviceFile. and i think it should live in a convenience module, either newscale.convenience or newscale.products or just newscale (top level). i don't see any reason why the 99% person should have to learn the stage/interface/multistage abstraction if they don't need to.

Cool; this is basically what's already implemented. 
````python
from newscale.stages import USBXYZStage

stage = USBXYZStage('COM4')
````
The above is really all you need to start moving stuff around. See the usb_multistage_test.py file for a demo.

Reply:
OK, but:
* I still think this should be in a different module than M3LinearSmartStage; it's quite a different concept. What if instead we made a new module called "multistage" and put the MultiStage and USBXYZStage (which derives from MultiStage) in there?
* While your example constructor USBXYZStage('COM4') is obvious enough, I still think the argument list for USBXYZStage.__init__() is confusing: Why is the baudrate an (optional) argument? Why is usb_interface still an argument? I understand that you're trying to enable lots of use cases but the point of this convenience class is to value simplicity over generality; there should be "one way" to instantiate and use this class. And, however the constructor works, it should be explained in the docstring, preferably with an example call.

Reply:
Ok, I just pulled all the multistage and convenience stuff into its own module ``multistage.py`` and left M3LinearSmartStage in ``stage.py``

Re 2: Ahh, it looks like the USBXYZStage was missing an example docstring.
To summarize:
* all options are laid out in the docstring now with the simplest at the top
* supporting the option of passing an interface in separately is great for simulations and mocking hardware

Reply:
Just tested and pushed these changes. I am marking as resolved.